runtime/lua/vim/lsp/snippet.lua

```diff
@@ -0,0 +1,406 @@
+    option = values[6],
+  }
+end)
+
+S.tabstop = P.any(
+  P.map(P.seq(S.dollar, S.int), function(values)
+    return {
+      type = 'tabstop',
+      tabstop = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.close), function(values)
+    return {
+      type = 'tabstop',
+      tabstop = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.transform, S.close), function(values)
+    return {
+      type = 'tabstop',
+      tabstop = values[3],
+      transform = values[4]
+    }
+  end)
+)
+
+S.placeholder = P.any(
+  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = 'placeholder',
+      tabstop = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.choice = P.map(P.seq(
+  S.dollar,
+  S.open,
+  S.int,
+  S.pipe,
+  P.many(
+    P.map(P.seq(S.text({ ',', '|' }), P.opt(S.comma)), function(values)
+      return values[1].esc
+    end)
+  ),
+  S.pipe,
+  S.close
+), function(values)
+  return {
+    type = 'choice',
+    tabstop = values[3],
+    items = values[5],
+  }
+end)
+
+S.variable = P.any(
+  P.map(P.seq(S.dollar, S.var), function(values)
+    return {
+      type = 'variable',
+      name = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.close), function(values)
+    return {
+      type = 'variable',
+      name = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.transform, S.close), function(values)
+    return {
+      type = 'variable',
+      name = values[3],
+      transform = values[4],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = 'variable',
+      name = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.snippet = P.map(P.many(P.any(S.toplevel, S.text({ '$' }, { '}', '\\' }))), function(values)
+  return {
+    type = 'snippet',
+    children = values,
+  }
+end)
+
+local M = {}
+
+---Parse snippet string and returns the AST
+---@param input string
+---@return table
+M.parse = function(input)
+  local result = S.snippet(input, 1)
+  if not result.parsed then
+    error('snippet parsing failed.')
+  end
+  return result.value
+end
+
+---Convert snippet nodes to string.
+---@param node table The result of M.parse
+---@param opts table|nil
+---  - with_placeholder boolean
+---  - resolve_variable fun(name: string): string
+---@return string
+M.to_string = function(node, opts)
+  opts = opts or {}
+
+  vim.validate({
+    node = { node, 't', 'node must be table.' },
+    opts = { opts, 't', 'opts must be table.' }
+  })
+
+  local to_string
+  to_string = function(node_)
+    local insert_text = {}
+    if node_.type == 'snippet' then
+      for _, c in ipairs(node_.children) do
+        table.insert(insert_text, to_string(c))
+      end
+    elseif node_.type == 'variable' then
```

Comment:
Can you pull node types out into an enum and validate against it instead of matching on strings? Thanks!

Reply:
I've added the `require('vim.lsp._snippet').NodeType` constants. (I'm refering `vim.lsp.protocol` to implement it